base/strings/search.jl

```diff
@@ -212,44 +206,50 @@ julia> rsearch("aaabbb","b")
 6:6
 ```
 """
-function rsearch(s::AbstractString, c::Chars, i::Integer)
-    e = endof(s)
-    j = search(RevString(s), c, e-i+1)
-    j == 0 && return 0
-    e-j+1
+function rsearch(s::AbstractString, c::Chars, i::Integer=start(s))
+    @boundscheck checkbounds(s, i)
+    isempty(c) && return i
+    j = i
+    while j > 0
+        d = s[j]
+        d in c && return j
+        j = prevind(s, j)
+    end
+    return 0
 end
 
 function _rsearchindex(s, t, i)
     if isempty(t)
-        return 1 <= i <= nextind(s,endof(s)) ? i :
-               throw(BoundsError(s, i))
+        @boundscheck checkbounds(s, i)
+        return i
     end
-    t = RevString(t)
-    rs = RevString(s)
     l = endof(s)
-    t1, j2 = next(t,start(t))
+    t1, j2 = last(t), endof(t)
     while true
-        i = rsearch(s,t1,i)
-        if i == 0 return 0 end
-        c, ii = next(rs,l-i+1)
-        j = j2; k = ii
+        i = rsearch(s, t1, i)
+        i == 0 && return 0
+        c = s[i]
+        ii = prevind(s, i)
+        j, k = j2, ii
         matched = true
-        while !done(t,j)
-            if done(rs,k)
+        while j > 0
+            if k < 1
                 matched = false
                 break
             end
-            c, k = next(rs,k)
-            d, j = next(t,j)
+            c, k = let rs_k = reverseind(s, k)
```

Comment:
`let` isn't really useful here, is it?

I also wonder whether both `reverseind` and `prevind` are needed here. Why do you need to work with indices in the reversed string?